Deprecate the world.opt build target and change default target to all

dra27 opened the pull request:
This builds upon @shindere's fine change in #8951 and the discussion in https://github.com/ocaml/ocaml/pull/8951#discussion_r328979928 to change the default target name to be `all`.

This PR:
- Changes the old `all` target to be `byteall` (which is technically more descriptive of what it actually does)
- Changes the default target name to be `all`
- Changes the behaviour of the default target to do `world`/`world.opt` as appropriate when first run, but then `byteall` thereafter (so no developer should ever have to run `make byteall`)
- Displays a deprecation warning at the end of the build if `make world` or `make world.opt` is used

The important observations:
- For core developers who need to care about the meaning of `make all`, its meaning is virtually unchanged (because one would expect such a developer to have already initialised `boot/` with a runtime and stdlib)
- Deprecating `world` and `world.opt` should encourage less surprising core developer practices when making changes which require bootstrapping (because you now don't end up writing a stdlib which requires a new feature "by accident").

The less important observations:
- We get to follow the GNU coding guidelines (whoopie!)
- Not all systems are sudo based... I quite like being able to do `make all install` occasionally, and it's less verbose and more normal than `make defaultentry install` (still a bit whoopie - doesn't work with `-j`, etc.)
- Especially in CI scripts, etc., I prefer an explicit target to be built rather than just `make` (I know, I'm weird) and again `make all` is clearer than `make defaultentry`

cc @xavierleroy

dra27 committed:
Default Makefile target is now all

The old all Makefile target is now technically byteall, but in practice
this distinction is irrelevant. make all now does what is always wanted:
it builds the system from cold if there is no runtime and stdlib in
boot/ and otherwise it behaves as the old all target used to.

In order to encourage better developer use of the targets, world and
world.opt are now deprecated: so development of a compiler which will
require bootstrapping should now be done less unknowingly.

gasche commented:
I'm a bit concerned with the change, especially the part where we deprecate things that are in the muscle memory of most contributors, without a very clear transition documentation (except this PR maybe, but it has a lot of implicit context.). (The deprecation message sounds like `make` replaces `make world`, and `make all` replaces `make world.opt`.)

To me "byteall" suggests "building everything needed for bytecode compilation". If it's not what it does, then it's probably not much clearer than "all".

I often decide between using `core`, `world` and `world.opt` depending on the things that I want to rebuild and be available. (The main thing that forces me to run `world.opt` is that the testsuite breaks in annoying way when a stale copy of the .opt world lies around; thanks to whoever is working on BYTECODE_ONLY again.) (The bootstrap-related subtleties were indeed lost on me, but I need to bootstrap so rarely that I don't care much.)

What's the approach now (after this PR) if I want to do `world` instead of `world.opt` on my machine, or the other way around?

xavierleroy commented:
I explained the reasons for the current choice of names here: https://github.com/ocaml/ocaml/pull/8951#discussion_r329591895

xavierleroy commented:
If we really must rename targets to satisfy @dra27, could we at least keep some semblance of symmetry between the "build the bytecode system" target (`byteall` in this PR, `all` previously) and "build the native-code system" target (`allopt`) ?  I mean, `byteall` is not symmetrical with `allopt`.

dra27 committed:
Rename byteall to allbyte

dra27 committed:
Clearer instructions in BOOTSTRAP.adoc

dra27 committed:
Keep world, but only coldstart once

dra27 committed:
Clearer deprecation message on world.opt

dra27 commented:
Thanks - I do appreciate your time on one of my foibles! `byteall` did mean what @gasche describes - the symmetry I'd gone with was `coreall` (`allopt` isn't available at the toplevel, where it's still `opt` for legacy reasons, presumably). However, `allbyte` is certainly a better symmetry - perhaps we could should add `allopt` at the toplevel just as an alias of `opt`? I do reiterate that you'd never need to use it - if `make all` would have run correctly before, it will still do that now.

With the "never have to run `make allbyte`" in mind, and also @gasche's point about muscle memory, I've un-deprecated `make world` (this removes instances of `make allbyte` from the CI scripts as well, which is pleasing). However, I've altered it so that it only does `coldstart` the first time. If renaming `defaulttarget` to `all` is my foible, improving the discipline in the use of the build system targets I think is a genuine improvement - this way `make world` does what does the core developer almost certainly *meant* (i.e. `boot/` is not touched, unless the tree is clean).

I have kept `world.opt` deprecated, but clarified the error message and also mentioned the fact that `make opt.opt` will rebuild the `.opt` compilers.

gadmm commented on BOOTSTRAP.adoc:
```diff
@@ -28,8 +28,9 @@ Here is how to perform a change that requires a bootstrap:
-state but builds more things than actually required.)
+safer. Similarly, `make all` will also bring you to such a stable
+state but builds more things than actually required, if the native compiler
+enabled.)
```
"is enabled"

dra27 replied:
Oops - thanks

gadmm commented on Makefile:
```diff
@@ -463,17 +474,21 @@ endif
+	@echo
+	@echo "$(MAKE) world.opt is deprecated; use $(MAKE) all \
+or even just $(MAKE)"
+	@echo "To rebuild the .opt tools, use $(MAKE) opt.opt"
```
As a relative novice of the compiler's build system, this is the sort of message that would annoy me: "why are they breaking my workflow now?". I am probably guilty of using `make world.opt` out of habit without understanding the subtleties, and I would be happy to learn that I have been doing something wrong. So the message could be the occasion to be more pedagogical, in fact it could be a simple advice where the word "deprecated" doesn't even need to appear (I find it a bit patronizing: "we have decided for you, and it is wrong not to do as we say").

dra27 replied:
I'm not sure which version of "now" you mean here - it's not being broken _now_, it's telling you that it may be broken in the future! The word deprecated is used pervasively - I feel you may need to find it less patronizing, if you'll inevitable excuse the meta-pun in that!

I softened the message slightly by re-wording "_you can now_ use".

gadmm replied:
Thank you for the rephrasing which is a bit better. It does not really address the point I was trying to make though. Now I am even more confused, even after reading again the full PR and discussions: first you say that the deprecation "should encourage less surprising core developer practices" (which? not sure I understood), but now you say that it may be broken in the future (how?). I am a user who does not have stakes in world.opt, and I am open-minded about learning better practices, but for now, I am not convinced by the the change. If you want to deprecate something, you first should be able to communicate clearly to users why it is being deprecated. (I do not think I understood what you meant in the second sentence either.)

dra27 replied:
I’m sorry for the confusion here, but I haven’t changed anything - deprecating means stop using it because it might not be there at some point in the future. I have no idea if/when that would actually happen.

dra27 replied:
As to why it's an improvement, the fact you don't know why sort of makes the point. At present, repeatedly running `make world.opt` will put a new runtime and standard library in `boot/` (that's the job of the `coldstart` target). If you have made changes which will ultimately require a bootstrap, you've now masked that. If you run the new `make all`, you'll know because your compiler won't build with the *original* one in `boot/` which was assembled the first time you ran `make all`.

gadmm replied:
You have no idea if/when the target world.opt would be changed or removed, so you are literally telling us that it is not deprecated.

Thanks for the explanation. I was already trusting that it is an improvement, otherwise you would not be doing it. But you have only just explained it to me now.

I repeat my advice of providing a pedagogical message instead. In fact, I think you could get your message across to users simply by providing a (usefully!) informative message, along the following lines:

* When initializing for the first time: "boot/ has been initialized; you can now use make all (or even just make) to compile without rebuilding it. To rebuild the .opt tools, use make opt.opt."
* When rebuilding boot/: "boot/ has been rebuilt."

gadmm replied:
Such an informative message could be appropriate for make world too.

gadmm reviewed (commented):
Advice about how to go with "deprecating"

dra27 committed:
Corrections

dra27 commented:
@gasche - I hope I've addressed all your points?

~@xavierleroy - have I persuaded you that `make all` continues to mean the same thing when developing (and you'll never have to run `make allbyte`?)~ (this isn't true)

gasche commented:
I'm not sure because I don't follow everything that is going on in this PR, but thanks for the effort! Here are some questions:

- I'm always configuring my machine with native-code support, but sometimes I want to run `world` or `world opt` (to quickly check that my changes are not breaking stuff) and other times `world.opt` (for benchmarking the resulting compiler, for example). What are the commands I should run after this PR? My understanding is that the answer is: `world` is now `allbyte`, `world opt` is `allbyte opt`, and `world.opt` is `opt.opt`. (I'm assuming that this is a non-bootstrap-requiring change so the `coldstart` part is irrelevant.) Is this mapping clearly documented somewhere for old people needing time to change their habits? (I don't see it in the PR.) Ideally the deprecation message for the world* targets could contain this mapping.

- Bonus sub-question: would there be a way to make this mapping more consistent? You've suggested defining an alias `allopt` for `opt`. Note that in #1821 I made sure that all sub-Makefiles consistently offered both an `allopt` and an `opt.opt` target (I was tired of having to guess which target was available in which subdirectory), and I made them synonyms to each other (maybe they should mean different things?). This would suggest having `allopt` rather be a synonym for `opt.opt`. (Nitpick: `opt.opt` is unclear as a name I think, it suggests that it builds the native version of the native compiler, which in fact it builds both the byte and native versions of both the byte and native compiler.)

- (new question) If I build in a given branch, and then I switch to a different branch (possibly a completely different OCaml version) and build again, is there a risk of the old-branch `boot/initialized` affecting my build in unexpected ways? I don't see any dependency in its Makefile rule, so the way I read the PR `coldstart` would not be run again, which seems wrong. (Something I didn't have to care about with my pedestrial usage of world* targets.) It's not immediately clear to me that we *can* have a system that is robust to branch changes (important to me) while also not populating boot at each build changing the compiler in a way that makes bootstrapping changes require more care (not important to me).

dra27 commented:
`world` is now unchanged (I altered the title) so you can still you do it. You *probably* should have been running `make allbyte opt` before which you can still do.

dra27 committed:
Add some info on allbyte to HACKING.adoc

dra27 commented:
Hmm, `allopt` is inconsistent in the tree - for tools it means build the `.opt` version (so part of `opt.opt` at the toplevel) and for libraries it means build the native code version of the library. So at the toplevel, it makes sense that `allopt` would be a synonym of `opt` to go with `allbyte` but then it's inconsistent with the meaning of `allopt` in all the other makefiles 😕

dra27 commented:
(I agree `opt.opt` is not necessarily the clearest name, although I can see how it means "with `opt` now build the `.opt` versions of everything")

xavierleroy commented:
I move to close this PR.  We've collectively wasted too much time on it already.  There's multicore OCaml to merge and a typechecker to rewrite, yet we're arguing about the names of Makefile entries.  Enough of this.
